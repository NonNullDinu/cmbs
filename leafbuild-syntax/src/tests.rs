fn parses_without_errors(s: &str) {
    assert_eq!(crate::parser::parse(s).errors, vec![]);
}

// Those are only *parsing* tests, so as long as they make sense
// *syntactically*, they should parse.
// Even if semantically they are void of meaning.
#[test]
fn unit_to_unit_assignment() {
    parses_without_errors("() = ()\n")
}

#[test]
fn line_trailing_space() {
    parses_without_errors("() = () \n")
}

#[test]
fn if_condition() {
    parses_without_errors("if (()) {} else {}")
}

#[test]
fn if_else_if_condition() {
    parses_without_errors("if () {} else if () {} else {}")
}

#[test]
fn if_else_if_else_condition_stretched() {
    parses_without_errors(
        r#"
    if ()

    {



    }

    else if () {


    }

    else {


    }
    "#,
    )
}

#[test]
fn if_condition_in_expr() {
    parses_without_errors("() = if () {} else {} \n")
}

#[test]
fn if_condition_in_expr_stretched() {
    parses_without_errors(
        r#"
    () = if ()
    {

    }

    else
    {


    }
    "#,
    )
}

#[test]
fn declaration() {
    parses_without_errors("let x = ()\n")
}

#[test]
fn foreach_basic() {
    parses_without_errors("foreach () in () {}")
}
