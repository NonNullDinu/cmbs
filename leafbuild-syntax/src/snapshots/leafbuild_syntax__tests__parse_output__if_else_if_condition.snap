---
source: leafbuild-syntax/src/tests.rs
expression: "if () {} else if () {} else {}"

---
- ROOT 0..30
  - Conditional 0..30
    - ConditionalBranch 0..8
      - "if" IF_KW 0..2
      - " " WHITESPACE 2..3
      - Expr 3..5
        - PrimaryExpr 3..5
          - TupleExpr 3..5
            - "(" L_PAREN 3..4
            - ")" R_PAREN 4..5
      - " " WHITESPACE 5..6
      - ExprBlock 6..8
        - "{" L_BRACE 6..7
        - "}" R_BRACE 7..8
    - " " WHITESPACE 8..9
    - "else" ELSE_KW 9..13
    - " " WHITESPACE 13..14
    - ConditionalBranch 14..22
      - "if" IF_KW 14..16
      - " " WHITESPACE 16..17
      - Expr 17..19
        - PrimaryExpr 17..19
          - TupleExpr 17..19
            - "(" L_PAREN 17..18
            - ")" R_PAREN 18..19
      - " " WHITESPACE 19..20
      - ExprBlock 20..22
        - "{" L_BRACE 20..21
        - "}" R_BRACE 21..22
    - " " WHITESPACE 22..23
    - "else" ELSE_KW 23..27
    - " " WHITESPACE 27..28
    - ExprBlock 28..30
      - "{" L_BRACE 28..29
      - "}" R_BRACE 29..30

